src/Common/FST.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+#include <vector>
+#include <string>
+#include <unordered_set>
+#include <unordered_map>
+#include <map>
+#include <memory>
+#include <array>
+#include <Core/Types.h>
+#include <base/types.h>
+#include <IO/WriteBuffer.h>
+
+namespace DB
+{
+using Output = UInt64;
+
+struct State;
+using StatePtr = std::shared_ptr<State>;
+
+struct Arc
+{
+    Arc() = default;
+    Arc(Output output_, const StatePtr &target_): output{output_}, target{target_}{}
+    Output output{0};
+    StatePtr target;
+
+    UInt16 serialize(WriteBuffer& write_buffer) const;
+};
+
+class ArcsBitmap
+{
+public:
+    void addArc(char label);
+    bool hasArc(char label) const;
+    int getIndex(char label) const;
+    int getArcNum() const;
+private:
+    friend struct State;
+    friend class FST;
+    UInt256 data{0};
+};
+
+struct State
+{
+    static constexpr size_t MAX_ARCS_IN_SEQUENTIAL_METHID = 32;
+    enum EncodingMethod
+    {
+        ENCODING_METHOD_SEQUENTIAL = 0,
+        ENCODING_METHOD_BITMAP,
+    };
+    State() = default;
+    State(const State &state) = default;
+
+    UInt64 hash() const;
+
+    Arc* getArc(char label);
+    void addArc(char label, Output output, StatePtr target);
+
+    void clear()
+    {
+        id = 0;
+        state_index = 0;
+        flag = 0;
+
+        arcs.clear();
+    }
+
+    UInt64 serialize(WriteBuffer& write_buffer);
+
+    UInt64 id{ 0 };
+    UInt64 state_index{ 0 };
+    struct Flag_Values
+    {
+        unsigned int is_final : 1;
+        unsigned int encoding_method : 2;
+    };
+
+    union
+    {
+        Flag_Values flag_values;
+        uint8_t flag{0};
+    };
+    std::map<char, Arc> arcs;
```

Comment:
Good comment. Using std::map in this case is overkill. Definitely it will be in our optimization list, but for now we changed it to std::unordered_map: according to[dont_need_map](https://cpp-optimizations.netlify.app/dont_need_map/), it is a reasonable solution.